### PR TITLE
Milestone 2 [School Library]: Decorate a Class

### PR DESCRIPTION
**Pull request by cherelemma**
# Implementation Summary

In this second milestone PR, I implemented the following project requirements. 

#### Interface
- Create a class `Nameable`.
   - Implement a method called `correct_name` that will raise a `NotImplementedError`.

#### Turn your Person class to Nameable
- Make sure that your Person class inherits from Nameable
- Make sure that this class has a method `correct_name` implemented. It should simply return the name attribute.

#### Prepare base Decorator
- Make sure that it inherits from Nameable.
- In the constructor assign a nameable object from params to an instance variable.
- Implement the `correct_name` method that returns the result of the correct_name method of the `@nameable`.

#### Prepare CapitalizeDecorator and TrimmerDecorator
- For the CapitalizeDecorator:
    - Create a class that inherits from the base Decorator class.
    - Implement a method `correct_name` that capitalizes the output of `@nameable.correct_name`.
- For the TrimmerDecorator:
    - Create a class that inherits from the base Decorator class.
    - Implement a method `correct_name` that makes sure that the output of `@nameable.correct_name` has a maximum of 10 characters. If it's longer it should trim the word.

### General Requirements

- [No linter errors](https://github.com/microverseinc/linters-config)
- Follow correct [GitFlow](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/git-github/articles/gitflow.md)
- Follow the list of [best practices for Ruby](https://github.com/microverseinc/curriculum-ruby/blob/main/articles/ruby_best_practices.md).

### CODE REVIEW

File: nameable.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class Nameable
+  def correct_name
+    puts 'NotImplementedError'
+  end
```
Comment (thecodechaser) [inline nameable.rb]: - [ ] Please make sure to raise the error and not to hard print it as requested in the project requirements.
  - You can consider this `raise NotImplementedError`.

File: base_decorator.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
+require_relative './nameable'
+
+class BaseDecorator < Nameable
+  def initialize(nameable)
+    super()
+    @nameable = nameable
+  end
+
```
Comment (thecodechaser) [inline base_decorator.rb]: - [ ] Please make sure the `nameable` variable is readable and writable so decorate pattern works as expected.
   - You can consider using this `attr_accessor :nameable`.

File: capitalize_decorator.rb [hunk 3]
```diff
@@ -0,0 +1,7 @@
+require_relative './base_decorator'
+
+class CapitalizeDecorator < BaseDecorator
+  def correct_name
+    @nameable.correct_name.upcase
+  end
```
Comment (thecodechaser) [inline capitalize_decorator.rb]: - [ ] Please make sure to capitalize the name as requested in the project requirements. The `upcase` method will not capitalize but will convert the entire string to uppercase.
  - You can consider using this `@nameable.correct_name.capitalize`.

**Review by thecodechaser — CHANGES_REQUESTED** [review]
## STATUS: CHANGES REQUIRED ♻️ 

## Hi @cherelemma👋, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## To Highlight 🎉
- No linter errors. :heavy_check_mark:
- Gitflow is followed. :heavy_check_mark:
- Professional README file. :heavy_check_mark:
- PR has a good title and summary. ✔️



## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by BregornOriginal — APPROVED** [review]
Hi @cherelemma 

Your project is complete!  There is nothing else to say other than... it's time to merge it 

Congratulations! 

giphy (2)

## Highlights

✅ Linters are passing
✅ Good commit history
✅ Good work implementing changes requested by the previous reviewer 💯 

## Optional suggestions
Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application.
Cheers and Happy coding!:aplauso::aplauso::aplauso:
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

Please, remember to tag me in your question so I can receive the notification.
------
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)[.](https://airtable.com/shrQAqnBwek5a0O0s)._)

**Pull request merged**